typed/compare.go

```diff
@@ -242,31 +253,67 @@ func (w *compareWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
 			// this element.
 			continue
 		}
-		// Ignore repeated occurences of `pe`.
-		if _, found := rValues.Get(pe); found {
-			continue
+		if v, found := rValues.Get(pe); found {
+			list := v.Unstructured().([]value.Value)
+			rValues.Insert(pe, value.NewValueInterface(append(list, rValue)))
+		} else {
+			rValues.Insert(pe, value.NewValueInterface([]value.Value{rValue}))
+			if _, found := lValues.Get(pe); !found {
+				allPEs = append(allPEs, pe)
+			}
 		}
-		rValues.Insert(pe, rValue)
-		// We can compare with nil if lValue is not present.
-		lValue, _ := lValues.Get(pe)
-		errs = append(errs, w.compareListItem(t, pe, lValue, rValue)...)
 	}
 
-	// Add items from left that are not in right.
-	for i := 0; i < lLen; i++ {
-		lValue := lhs.At(i)
-		pe, err := listItemToPathElement(w.allocator, w.schema, t, lValue)
-		if err != nil {
-			errs = append(errs, errorf("element %v: %v", i, err.Error())...)
-			// If we can't construct the path element, we can't
-			// even report errors deeper in the schema, so bail on
-			// this element.
-			continue
+	for _, pe := range allPEs {
+		lList := []value.Value(nil)
+		if l, ok := lValues.Get(pe); ok {
+			lList = l.Unstructured().([]value.Value)
 		}
-		if _, found := rValues.Get(pe); found {
-			continue
+		rList := []value.Value(nil)
+		if l, ok := rValues.Get(pe); ok {
+			rList = l.Unstructured().([]value.Value)
+		}
+
+		switch {
+		case len(lList) == 0 && len(rList) == 0:
+			// We shouldn't be here anyway.
+			return
+		case len(lList) <= 1 && len(rList) <= 1:
+			lValue := value.Value(nil)
+			if len(lList) != 0 {
+				lValue = lList[0]
+			}
+			rValue := value.Value(nil)
+			if len(rList) != 0 {
+				rValue = rList[0]
+			}
+			errs = append(errs, w.compareListItem(t, pe, lValue, rValue)...)
+		case len(lList) >= 2 && len(rList) >= 2:
+			listEqual := func(lList, rList []value.Value) bool {
+				if len(lList) != len(rList) {
+					return false
+				}
+				for i := range lList {
+					if !value.Equals(lList[i], rList[i]) {
+						return false
+					}
+				}
+				return true
+			}
+			if !listEqual(lList, rList) {
+				w.comparison.Modified.Insert(append(w.path, pe))
+			}
```

Comment:
If I'm reading this correctly, this is the "no duplicates" case and should be semantically identical to the old code?